### ISSUE TITLE
`--n_cores` could detect the number of cores automatically

### ISSUE THREAD

**Issue by beurdouche** [FStarLang/FStar#922]
Something that would be very nice to have is to have F* detect the max number of cores on the machine automatically when `--n_cores` with no argument is provided :)

It would make our Makefiles portable and avoid us to hardcode a specific number

Follow up of #146

**Comment by catalin-hritcu** [comment]
There seem to be many ways to do this, but which one is portable. The following work on my machine
```
$ grep -c ^processor /proc/cpuinfo

8
[hritcu@resurrected FStar]$ nproc
8
[hritcu@resurrected FStar]$ getconf _NPROCESSORS_ONLN
8
```
https://stackoverflow.com/questions/6481005/how-to-obtain-the-number-of-cpus-cores-in-linux-from-the-command-line

**Comment by catalin-hritcu** [comment]
And once you know the number of cores, to what would you set --n_cores? I don't think it's reasonable to assume that you can take over *all* the cores of a machine. What about half?

**Comment by shaolintl** [comment]
maybe you can put a 0..1 value, i.e 0.5? and call the flag --n_cores_per?

**Comment by beurdouche** [comment]
I would say that the default could be `max - 1` or eventually `max - 2` to be conservative.
It's not like people don't know that F* has a performance cost :)

**Comment by kyoDralliam** [comment]
Isn't it the kind of parameter for which you *do* want to use `OTHERFLAGS` ? I don't want to use all my cores for one F* task when I have 2 other tasks running in parallel.

**Comment by beurdouche** [comment]
Yes, you are right. But we can't hardcode `--n_cores 72` in our makefiles. Maybe keeping the default to what it is now and having `--n_cores max` that actually uses (max number of cores - 1) could be ok ?

**Comment by catalin-hritcu** [comment]
To be honest I don't see us using `--n_cores` in our default makefiles, since `make -j` is so much better for reliably getting good CPU utilization for this kind of batch processing. The question still applies for the argument to `make -j` and in case we want to pass `--n_cores` by default in interactive mode.

**Comment by A-Manning** [comment]
`System.Environment.ProcessorCount` is probably what you want for this, as this works on both windows and mono

**Comment by catalin-hritcu** [comment]
Here's Kenji and my thinking on this: we don't want --n_cores to be passed by the default, and for just figuring out the number of cores on the machine there are various solutions in this thread, which one can use locally. Closing.
